templates/tfengine/components/terraform/main/main.tf

```diff
@@ -12,12 +12,36 @@
 # See the License for the specific language governing permissions and
 # limitations under the License.
 
+{{- $hasGoogle := false}}
+{{- $hasGoogleBeta := false}}
+{{- $hasKubernetes := false}}
+
 terraform {
   required_version = ">=1.0"
   required_providers {
+{{if has . "providers" -}}
+  {{range .providers -}}
+    {{.name}} = "{{.version_constraints}}"
+    {{if or (eq .name "google") (eq .name "hashicorp/google") -}}
+      {{$hasGoogle = true -}}
+    {{end -}}
+    {{if or (eq .name "google-beta") (eq .name "hashicorp/google-beta") -}}
+      {{$hasGoogleBeta = true -}}
+    {{end -}}
+    {{if or (eq .name "kubernetes") (eq .name "hashicorp/kubernetes") -}}
+      {{$hasKubernetes = true -}}
+    {{end -}}
+  {{end -}}
+{{end -}}
+    {{if not $hasGoogle -}}
```

Comment:
Why is this block necessary?

Reply:
It will complain if multiple constraints for the same provider is specified, and I still want to set the same base provider constraints for google, google-beta and kubernetes providers as what we do now.